persim/pers_landscape_approx.py

```diff
@@ -0,0 +1,380 @@
+            print("Bad choice of grid, values is empty")
+
+        self.values = L
+        return
+
+    def values_to_pairs(self):
+        """Converts function values to ordered pairs and returns them."""
+        self.compute_landscape()
+        grid_values = list(np.linspace(self.start, self.stop, self.num_steps))
+        result = []
+        for vals in self.values:
+            pairs = list(zip(grid_values, vals))
+            result.append(pairs)
+        return np.array(result)
+
+    def __add__(self, other):
+        super().__add__(other)
+        if self.start != other.start:
+            raise ValueError("Start values of grids do not coincide")
+        if self.stop != other.stop:
+            raise ValueError("Stop values of grids do not coincide")
+        if self.num_steps != other.num_steps:
+            raise ValueError("Number of steps of grids do not coincide")
+        self_pad, other_pad = union_vals(self.values, other.values)
+        return PersLandscapeApprox(
+            start=self.start,
+            stop=self.stop,
+            num_steps=self.num_steps,
+            hom_deg=self.hom_deg,
+            values=self_pad + other_pad,
+        )
+
+    def __neg__(self):
+        return PersLandscapeApprox(
+            start=self.start,
+            stop=self.stop,
+            num_steps=self.num_steps,
+            hom_deg=self.hom_deg,
+            values=np.array([-1 * depth_array for depth_array in self.values]),
+        )
+        pass
+
+    def __sub__(self, other):
+        return self + -other
+
+    def __mul__(self, other: float):
+        super().__mul__(other)
+        return PersLandscapeApprox(
+            start=self.start,
+            stop=self.stop,
+            num_steps=self.num_steps,
+            hom_deg=self.hom_deg,
+            values=np.array([other * depth_array for depth_array in self.values]),
+        )
+
+    def __rmul__(self, other: float):
+        return self.__mul__(other)
+
+    def __truediv__(self, other: float):
+        super().__truediv__(other)
+        return (1.0 / other) * self
+
+    def __getitem__(self, key: slice) -> list:
+        """
+        Returns a list of values corresponding in range specified by
+        depth
+
+        Parameters
+        ----------
+        key : slice object
+
+        Returns
+        -------
+        list
+            The values of the landscape function corresponding
+        to depths given by key
+        """
+        self.compute_landscape()
+        return self.values[key]
+
+    def p_norm(self, p: int = 2) -> float:
+        return np.sum([np.linalg.norm(depth, p) for depth in self.values])
+
+    def sup_norm(self) -> float:
+        return np.max(np.abs(self.values))
+
+
+############################################
+# End PersLandscapeApprox class definition #
+############################################
+
+
+def snap_PL(
+    pls: list, start: float = None, stop: float = None, num_steps: int = None
```

Comment:
I _love_ the typing by the way :D 

With this typing, can we do some static analysis of the code as well?  I'd love to incorporate that into our testing workflow if so.

Reply:
Thanks. I'm not an expert on this, but I think some tools (like [mypy](https://mypy.readthedocs.io/en/stable/introduction.html)) allow you to do just that.